util/common_copy_script.py

```diff
@@ -0,0 +1,180 @@
+            src_uri_without_scheme = source_path_str[len("gs://"):]
+            src_parts = src_uri_without_scheme.split('/', 1)
+            src_bucket_name = src_parts[0]
+            src_gcs_prefix = src_parts[1] if len(src_parts) > 1 else ""
+            if src_gcs_prefix and not src_gcs_prefix.endswith('/'):
+                src_gcs_prefix += '/'
+
+            try:
+            
+                source_bucket = gcs_client.bucket(src_bucket_name)
+                source_bucket.exists()
+            except Exception as e:
+                raise RuntimeError(f"Unexpected error initializing source bucket '{src_bucket_name}': {e}") from e
+            logging.info(f"GCS source bucket: '{src_bucket_name}', prefix: '{src_gcs_prefix}'")
+
+        if not is_target_gcs:
+            target_local_path = Path(target_path_str)
+            try:
+                target_local_path.mkdir(parents=True, exist_ok=True)
+                logging.info(f"Local target folder '{target_local_path}' ensured to exist.")
+            except OSError as e:
+                raise IOError(f"Error creating local target folder '{target_local_path}': {e}") from e
+        else:
+            # gcs path logic to chech that path is valid and fetched the bucket name and folder name
+            if not target_path_str.startswith("gs://"):
+                raise ValueError(f"Invalid GCS URI for target: '{target_path_str}'. Must start with 'gs://'.")
+            dest_uri_without_scheme = target_path_str[len("gs://"):]
+            dest_parts = dest_uri_without_scheme.split('/', 1)
+            dest_bucket_name = dest_parts[0]
+            dest_gcs_prefix = dest_parts[1] if len(dest_parts) > 1 else ""
+            if dest_gcs_prefix and not dest_gcs_prefix.endswith('/'):
+                dest_gcs_prefix += '/'
+
+            try:
+                target_bucket = gcs_client.bucket(dest_bucket_name)
+                target_bucket.exists() 
+            except Exception as e:
+                raise RuntimeError(f"Unexpected error initializing target bucket '{dest_bucket_name}': {e}") from e
+            logging.info(f"GCS target bucket: '{dest_bucket_name}', prefix: '{dest_gcs_prefix}'")
+
+
+        # Iterating the  source items and perform copy operation
+        if not is_source_gcs: 
+            logging.info(f"Iterating local source '{source_local_path}' (only files directly in this directory will be copied).")
+            for item in source_local_path.iterdir():
+                if not item.is_file():
+                    logging.info(f"  Skipped local non-file item: '{item.name}'")
+                    skipped_count += 1
+                    continue
+
+                source_item_name = item.name # The base filename from the local source
+                try:
+                     # Local to Local Copy operation
+                    if not is_target_gcs:
+                        destination_path = target_local_path / source_item_name
```

Comment:
you mean `target_local_path + '/' +  source_item_name` ?
A better way is to us os.path.join(target_local_path, source_item_name) that can handle trailing '/' in target_local_path